DOCSP-41865: datetime serializers (#173)
mongodb/docs-kotlin: source/fundamentals/data-formats/serialization.txt

      Retrieving as Document type
      Document{{_id=..., _t=Teacher, name=Vivian Lee, department=History}}
      Document{{_id=..., _t=Student, name=Kate Parker, grade=10}}

.. _kotlin-datetime-serialization:

Serialize Dates and Times
-------------------------

In this section, you can learn about using {+language+} serialization to
work with date and time types.

kotlinx-datetime Library
~~~~~~~~~~~~~~~~~~~~~~~~

> ``kotlinx-datetime`` is a {+language+} library that allows
> you to work with dates and times. If you want a high level of control
> over how your date and time values are serialized, you can add the
> dependency to your project's dependency list.

Reviewer: S: maybe a little more straightforward

```suggestion
``kotlinx-datetime`` is a {+language+} library that offers
a high level of control over how your date and time values
are serialized. To use the library, add the ``kotlinx-datetime``
dependency to your project's dependency list.
```

Select from the following tabs to see how to add the ``kotlinx-datetime``
dependency to your project by using the :guilabel:`Gradle` and
:guilabel:`Maven` package managers:

.. tabs::

   .. tab::
      :tabid: Gradle

      .. code-block:: kotlin
        :caption: build.gradle.kts

        implementation("org.jetbrains.kotlinx:kotlinx-datetime:{+kotlinx-dt-version+}")

   .. tab::
      :tabid: Maven

      .. code-block:: kotlin
        :caption: pom.xml

        <dependency>
            <groupId>org.jetbrains.kotlinx</groupId>
            <artifactId>kotlinx-datetime-jvm</artifactId>
            <version>{+kotlinx-dt-version+}</version>
        </dependency>

To learn more about this library, see the :github:`kotlinx-datetime repository
</Kotlin/kotlinx-datetime>` on GitHub.

Example Data Class with Dates and Times
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

After you add the library dependency, you can implement serializers from
the ``kotlinx-datetime`` library that map your data class field values
to the expected types in BSON.

> The following ``Appointment`` data class has the ``@Serializable``
> annotation and the ``@Contextual`` annotation on the ``date`` field
> to use the ``kotlinx-datetime`` serializer for ``LocalDate`` values.

Reviewer: S: i had a little trouble parsing this all at first. i tried a few versions and ended up with a list, but feel free to use/discard parts as you see fit. it also sort of bleeds into the following example about the time being stored as a string.

```suggestion
In the following code example, the driver serializes all fields in
the ``Appointment`` data class in the following ways:

 -  ``name``: The driver serializes the value as a string.
 - ``date``:  The driver uses the ``kotlinx-datetime`` serializer because
    the field includes the ``@Contextual`` annotation. The value is serialized
    as a BSON date.
 - ``time``: The driver serializes the value as a string.
```

> Values of the ``time`` field are stored as strings, because the driver

Reviewer: Q: Does this code example show that the `time` field is stored as a string?

Author: yes, this is the default behavior of the driver. The result is shown later in the section

the ``kotlinx-datetime`` library that map your data class field values
to the expected types in BSON.

In the following example, the driver serializes the fields of
the ``Appointment`` data class with the following behavior:

- ``name``: The driver serializes the value as a string.

- ``date``:  The driver uses the ``kotlinx-datetime`` serializer

>   because the field has the ``@Contextual`` annotation. So,
>   ``LocalDate`` values are serialized as BSON dates.

Reviewer: S:

```suggestion
  because the field has the ``@Contextual`` annotation.
  ``LocalDate`` values are serialized as BSON dates.
```
